db/upgrade.php

```diff
@@ -13,57 +13,84 @@
-
-        // Conditionally launch create table for tool_mergeusersr
+        // Conditionally launch create table for tool_mergeusers.
         if (!$dbman->table_exists($table)) {
             $dbman->create_table($table);
-        }
-
-        // mergeusers savepoint reached
+        }        
+        // Mergeusers savepoint reached.
         upgrade_plugin_savepoint(true, 2013112912, 'tool', 'mergeusers');
     }
-
+    if ($oldversion < 2023060801) {
+        // Define table tool_mergeusers_queue to be created.
+        $table = new xmldb_table('tool_mergeusers_queue');
+        // Adding fields to table tool_mergeusers_queue.
+        $table->add_field('id', XMLDB_TYPE_INTEGER, '10', null, XMLDB_NOTNULL, XMLDB_SEQUENCE, null);
+        $table->add_field('removeuserfield', XMLDB_TYPE_CHAR, '50', null, XMLDB_NOTNULL, null, null);
+        $table->add_field('removeuservalue', XMLDB_TYPE_CHAR, '255', null, XMLDB_NOTNULL, null, null);
+        $table->add_field('removeuserid', XMLDB_TYPE_INTEGER, '10', null, null, null, null);
+        $table->add_field('keepuserfield', XMLDB_TYPE_CHAR, '50', null, XMLDB_NOTNULL, null, null);
+        $table->add_field('keepuservalue', XMLDB_TYPE_CHAR, '255', null, XMLDB_NOTNULL, null, null);
+        $table->add_field('keepuserid', XMLDB_TYPE_INTEGER, '10', null, null, null, null);
+        $table->add_field('timeadded', XMLDB_TYPE_INTEGER, '10', null, XMLDB_NOTNULL, null, null);
+        $table->add_field('timemodified', XMLDB_TYPE_INTEGER, '10', null, null, null, null);
+        $table->add_field('status', XMLDB_TYPE_INTEGER, '4', null, XMLDB_NOTNULL, null, null);
+        $table->add_field('retries', XMLDB_TYPE_INTEGER, '4', null, XMLDB_NOTNULL, null, null);
+        $table->add_field('log', XMLDB_TYPE_TEXT, null, null, null, null, null);
+        // Adding keys to table tool_mergeusers_queue.
+        $table->add_key('primary', XMLDB_KEY_PRIMARY, ['id']);
+        // Adding indexes to table tool_mergeusers_queue.
+        $table->add_index('mdl_toolmergqueu_ruv_ix', XMLDB_INDEX_NOTUNIQUE, ['removeuservalue']);
+        $table->add_index('mdl_toolmergqueu_kuv_ix', XMLDB_INDEX_NOTUNIQUE, ['keepuservalue']);
+        $table->add_index('mdl_toolmergqueu_tas_ix', XMLDB_INDEX_NOTUNIQUE, ['timeadded', 'status']);
+        $table->add_index('mdl_toolmergqueu_tms_ix', XMLDB_INDEX_NOTUNIQUE, ['timemodified', 'status']);
+        $table->add_index('mdl_toolmergqueu_rui_ix', XMLDB_INDEX_NOTUNIQUE, ['removeuserid']);
+        $table->add_index('mdl_toolmergqueu_kui_ix', XMLDB_INDEX_NOTUNIQUE, ['keepuserid']);
+        // Conditionally launch create table for tool_mergeusers_queue and populate it importing data from old table.
+        if (!$dbman->table_exists($table)) {
+            $dbman->create_table($table);
+            // Export data from mergeusers table to the new one tool_mergeusers_queue.
+            merge_request::export_data_to_new_table();
```

Comment:
Could you put this line into a second upgrade step, so that it can be reprocessed properly in case of failure.

I propose something like, removing this line from here, and add a new section like:

```php
if ($oldversion < 2023060802) {
    merge_request::export_data_to_new_table();
    upgrade_plugin_savepoint(true, 2023060802, 'tool', 'mergeusers');
}
```

Reply:
update the values compared to $oldversion as you like, or update the version.php with this new version number.

Reply:
> > > 4. You also need to fix git conflicts expressed by github here.
> > 
> > 
> > which is the best way to fix git conflicts expressed by gituhb here? git merge or git rebase ?
> 
> At some time, I wrote a comment here that I built a branch with name `nvallinoto-master` in this same repo, with all conflicts being solved. And I suggested you to start from there with new changes. Have you proceed in this way as I suggested? This would be very benefitial for both nowadays. If not, it would be better to repeat conflict solving again from your branch. If you look to the nvallinoto-master branch and identify easily the commits there, maybe you can `git cherry-pick` there the rest of the commits.
> 
> In general, I prefer to make a `git rebase`, so that you put all your changes on top of the jpahullo/master branch, and solve any conflict appears during the process.

perhaps the best way at the end of this tour of commits is to create a new repository that take a new/fresh fork of the last commit of merge user plugin and then I add new files and commits to this new repository and then create new pull request from this to your [jpahullo:wip-ws-tasks](https://github.com/jpahullo/moodle-tool_mergeusers/tree/wip-ws-tasks)

Reply:
as I commented on the other comment, proceed in the way you feel more comfortable. Keep in mind not to loose changes! Please!